docs/community/20241112-node-bootstrapping.md

```diff
@@ -0,0 +1,118 @@
+---
+title: ClusterAPI Node Bootstrapping Working Group
+authors:
+    - "@t-lo"
+reviewers:
+    - "@elmiko"
+    - "@eljohnson92"
+    - "@johananl"
+    - "@tormath1"
+    - "@fabriziopandini"
+    - "@sbueringer"
+
+creation-date: 2024-11-12
+last-updated: 2024-11-15
+status: proposed
+see-also:
+  - https://github.com/kubernetes-sigs/cluster-api/issues/5294
+  - https://docs.google.com/document/d/1Fz5vWwhWA-d25_QDqep0LWF6ae0DnTqd5-8k8N0vDDM/edit
+  - https://github.com/kubernetes-sigs/cluster-api/issues/9157
+  - https://youtu.be/0AhA4Box3MM?si=IfKJfMWmlA9EW7ri&t=172
+---
+
+# Node Bootstrapping Working Group ("CAPI-NoBo")
+
+This document briefly outlines the scope, communication media, and
+stakeholders for a Working Group dedicated to the Provisioning aspect of
+Node Bootstrapping.
+
+Provisioning, in this context, refers to configuring and customising the
+node operating system to prepare it to serve as a ClusterAPI cluster node.
+
+
+## User Story and Problem Statement
+
+As a Linux Distribution maintainer aiming to support ClusterAPI, or to improve
+my existing support of ClusterAPI in my distribution, I would like to be able to
+cleanly integrate my contributions into the node bootstrapping process without
+interfering with other implementations.
+I would like my contributions to be as re-usable as possible across different
+bootstrap providers.
+I would like to read and to follow documentation, guidelines, and specifications
+on the above.
+I would like to offer a choice of node bootstrapping configuration systems to users,
+enable distribution vendors to extend the choices available by adding new
+configuration systems, and ease maintenance and enable extension of currently
+supported systems for the ClusterAPI developer community.
+
+
+**Problem statement / Example issues**
+
+Currently, the kubeadm bootstrap provider implements two systems to configure a node's
+operating system at provisoining time (referred to as "OS provisioning" from here on):
+cloud-init and ignition.
+As there is currently no abstract way to pass OS provisioning configuration to node
+bootstrapping, both OS provisioning system implementations - ignition and cloud-init - must cover
+both generic node bootstrapping as well as OS provisioning system specific concerns.
+This leads to duplicate code and to maintenance overhead.
+Also, other providers like e.g. MicroK8s must implement their own OS provisioning
+systems since kubeadm's implementations is specific to kubeadm and cannot be re-used.
+MicroK8s currently only supports cloud-init.
```

Comment:
```suggestion
As there is currently no OS provisioning configuration abstraction, the kubeadm bootstrap provider
is tightly coupled with cloud-init and Ignition. Furthermore, the Ignition implementation is built
on top of the cloud-init one. This makes it hard to develop provisioning implementations
independently and makes it hard to reuse code effectively between bootstrap implementations: Other
bootstrap providers such as MicroK8s must implement their own OS provisioning, and MicroK8s
currently only supports cloud-init.

In addition, the current design mixes bootstrap-related provisioning code (such as the commands
necessary for executing the kubeadm binary on a host) with generic OS customizations (such as disk
partitioning or Linux user configuration).
```

I suggest this rephrasing since IMO the current phrasing doesn't emphasize the extent of the current issues.

Our main problems:

- Bootstrap is tightly coupled with Provisioning. This is the biggest issue IMO.
- The Ignition implementation is built on top of the cloud-init implementation. This means that we not only have a tight coupling between bootstrap and provisioning, **in addition** Ignition fully depends on cloud-init.

The implications of these problems:

- We can't modify cloud-init or Ignition code without affecting kubeadm-related code or vice versa.
- We can't modify cloud-init code without affecting Ignition code.
- We can't flexibly combine bootstrap providers with provisioning implementations (regardless of the question of whether provisioning should be embedded inside the bootstrap provider or be fully independent).
- Provisioning config is "abused" by users who want to perform generic OS customization (unrelated to CAPI bootstrap, i.e. disk partitioning) and have no well-defined mechanism for doing so.

I've explained all of the above in detail under the [Motivation section](https://docs.google.com/document/d/1Fz5vWwhWA-d25_QDqep0LWF6ae0DnTqd5-8k8N0vDDM/edit?tab=t.0#heading=h.wcf1gbhz0yw1) of the WIP design proposal (I think it's relevant to read that section even if we don't end up adopting the specific proposed design).

Reply:
This is great! Thank you for elaborating.